File: Source/OldCode/Archive/2005 TVA Code Library/Version/2.0/Source/Tva.Phasors/CommandFrameBase.vb
'*******************************************************************************************************
'  CommandFrameBase.vb - Command frame base class
'  Copyright © 2005 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2005
'  Primary Developer: James R Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  01/14/2005 - James R Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

Imports System.Text
Imports Tva.Interop
Imports Tva.Phasors.Common

' This class represents the protocol independent common implementation of a command frame that can be sent or received from a PMU.
<CLSCompliant(False)> _
Public MustInherit Class CommandFrameBase

    Inherits ChannelFrameBase(Of ICommandCell)
    Implements ICommandFrame

    Private m_command As Command

    Protected Sub New(ByVal cells As CommandCellCollection, ByVal command As Command)

        MyBase.New(cells)
        m_command = command

    End Sub

    ' Derived classes are expected to expose a Public Sub New(ByVal binaryImage As Byte(), ByVal startIndex As Int32)
    ' and automatically pass in parsing state
    Protected Sub New(ByVal state As ICommandFrameParsingState, ByVal binaryImage As Byte(), ByVal startIndex As Int32)

        MyBase.New(state, binaryImage, startIndex)

    End Sub

    ' Derived classes are expected to expose a Public Sub New(ByVal commandFrame As ICommandFrame)
    Protected Sub New(ByVal commandFrame As ICommandFrame)

        MyClass.New(commandFrame.Cells, commandFrame.Command)

    End Sub

    Protected Overrides ReadOnly Property FundamentalFrameType() As FundamentalFrameType
        Get
            Return Phasors.FundamentalFrameType.CommandFrame
        End Get
    End Property

    Public Overridable Shadows ReadOnly Property Cells() As CommandCellCollection Implements ICommandFrame.Cells
        Get
            Return MyBase.Cells
        End Get
    End Property

    Public Overridable Property Command() As Command Implements ICommandFrame.Command
        Get
            Return m_command
        End Get
        Set(ByVal value As Command)
            m_command = value
        End Set
    End Property

    Public Overridable Property ExtendedData() As Byte() Implements ICommandFrame.ExtendedData
        Get
            Return Cells.BinaryImage
        End Get
        Set(ByVal value As Byte())
            Cells.Clear()
            MyBase.ParseBodyImage(New CommandFrameParsingState(Cells, 0, value.Length), value, 0)
        End Set
    End Property

    Protected Overrides ReadOnly Property BodyLength() As UInt16
        Get
            Return MyBase.BodyLength + 2
        End Get
    End Property

    Protected Overrides ReadOnly Property BodyImage() As Byte()
        Get
            Dim buffer As Byte() = Array.CreateInstance(GetType(Byte), BodyLength)
            Dim index As Int32 = 2

            EndianOrder.BigEndian.CopyBytes(m_command, buffer, 0)
            CopyImage(MyBase.BodyImage, buffer, index, MyBase.BodyLength)

            Return buffer
        End Get
    End Property

    Protected Overrides Sub ParseBodyImage(ByVal state As IChannelParsingState, ByVal binaryImage() As Byte, ByVal startIndex As Int32)

        m_command = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex)
        MyBase.ParseBodyImage(state, binaryImage, startIndex + 2)

    End Sub

End Class

File: Source/Libraries/GSF.PhasorProtocols/CommandFrameBase.cs
//******************************************************************************************************
//  CommandFrameBase.cs - Gbtc
//
//  Copyright © 2012, Grid Protection Alliance.  All Rights Reserved.
//
//  Licensed to the Grid Protection Alliance (GPA) under one or more contributor license agreements. See
//  the NOTICE file distributed with this work for additional information regarding copyright ownership.
//  The GPA licenses this file to you under the MIT License (MIT), the "License"; you may
//  not use this file except in compliance with the License. You may obtain a copy of the License at:
//
//      http://www.opensource.org/licenses/MIT
//
//  Unless agreed to in writing, the subject software distributed under the License is distributed on an
//  "AS-IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. Refer to the
//  License for the specific language governing permissions and limitations.
//
//  Code Modification History:
//  ----------------------------------------------------------------------------------------------------
//  01/14/2005 - J. Ritchie Carroll
//       Generated original version of source code.
//  09/15/2009 - Stephen C. Wills
//       Added new header and license agreement.
//  10/5/2012 - Gavin E. Holden
//       Added new header and license agreement.
//  12/17/2012 - Starlynn Danyelle Gilliam
//       Modified Header.
//
//******************************************************************************************************

using System;
using System.Collections.Generic;
using System.Runtime.Serialization;
using GSF.Parsing;

namespace GSF.PhasorProtocols
{
    /// <summary>
    /// Represents the protocol independent common implementation of any <see cref="ICommandFrame"/> that can be sent or received.
    /// </summary>
    [Serializable]
    public abstract class CommandFrameBase : ChannelFrameBase<ICommandCell>, ICommandFrame
    {
        #region [ Members ]

        // Fields
        private DeviceCommand m_command;

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="CommandFrameBase"/> from specified parameters.
        /// </summary>
        /// <param name="cells">The reference to the <see cref="CommandCellCollection"/> for this <see cref="CommandFrameBase"/>.</param>
        /// <param name="command">The <see cref="DeviceCommand"/> for this <see cref="CommandFrameBase"/>.</param>
        protected CommandFrameBase(CommandCellCollection cells, DeviceCommand command)
            : base(0, cells, 0)
        {
            m_command = command;
        }

        /// <summary>
        /// Creates a new <see cref="CommandFrameBase"/> from serialization parameters.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> with populated with data.</param>
        /// <param name="context">The source <see cref="StreamingContext"/> for this deserialization.</param>
        protected CommandFrameBase(SerializationInfo info, StreamingContext context)
            : base(info, context)
        {
            // Deserialize command frame
            m_command = (DeviceCommand)info.GetValue("command", typeof(DeviceCommand));
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets the <see cref="FundamentalFrameType"/> for this <see cref="CommandFrameBase"/>.
        /// </summary>
        public override FundamentalFrameType FrameType => FundamentalFrameType.CommandFrame;

        /// <summary>
        /// Gets reference to the <see cref="CommandCellCollection"/> for this <see cref="CommandFrameBase"/>.
        /// </summary>
        public new virtual CommandCellCollection Cells => base.Cells as CommandCellCollection;

        /// <summary>
        /// Gets or sets the parsing state for the this <see cref="CommandFrameBase"/>.
        /// </summary>
        public new virtual ICommandFrameParsingState State
        {
            get => base.State as ICommandFrameParsingState;
            set => base.State = value;
        }

        /// <summary>
        /// Gets or sets the <see cref="DeviceCommand"/> for this <see cref="CommandFrameBase"/>.
        /// </summary>
        public virtual DeviceCommand Command
        {
            get => m_command;
            set => m_command = value;
        }

        /// <summary>
        /// Gets or sets extended binary image data for this <see cref="CommandFrameBase"/>.
        /// </summary>
        public virtual byte[] ExtendedData
        {
            get => Cells.BinaryImage();
            set
            {
                Cells.Clear();
                State = new CommandFrameParsingState(0, value.Length, true, true);
                ParseBodyImage(value, 0, value.Length);
            }
        }

        /// <summary>
        /// Gets the length of the <see cref="BodyImage"/>.
        /// </summary>
        protected override int BodyLength => base.BodyLength + 2;

        /// <summary>
        /// Gets the binary body image of this <see cref="CommandFrameBase"/>.
        /// </summary>
        protected override byte[] BodyImage
        {
            get
            {
                byte[] buffer = new byte[BodyLength];
                int index = 2;

                BigEndian.CopyBytes((short)m_command, buffer, 0);
                base.BodyImage.CopyImage(buffer, ref index, base.BodyLength);

                return buffer;
            }
        }

        /// <summary>
        /// <see cref="Dictionary{TKey,TValue}"/> of string based property names and values for the <see cref="CommandFrameBase"/> object.
        /// </summary>
        public override Dictionary<string, string> Attributes
        {
            get
            {
                Dictionary<string, string> baseAttributes = base.Attributes;

                baseAttributes.Add("Device Command", (int)Command + ": " + Command);

                if (Cells.Count > 0)
                    baseAttributes.Add("Extended Data", ByteEncoding.Hexadecimal.GetString(Cells.BinaryImage()));
                else
                    baseAttributes.Add("Extended Data", "<null>");

                return baseAttributes;
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Parses the binary body image.
        /// </summary>
        /// <param name="buffer">Binary image to parse.</param>
        /// <param name="startIndex">Start index into <paramref name="buffer"/> to begin parsing.</param>
        /// <param name="length">Length of valid data within <paramref name="buffer"/>.</param>
        /// <returns>The length of the data that was parsed.</returns>
        protected override int ParseBodyImage(byte[] buffer, int startIndex, int length)
        {
            int parsedLength = 2;

            m_command = (DeviceCommand)BigEndian.ToInt16(buffer, startIndex);
            parsedLength += base.ParseBodyImage(buffer, startIndex + 2, length);

            return parsedLength;
        }

        /// <summary>
        /// Populates a <see cref="SerializationInfo"/> with the data needed to serialize the target object.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> to populate with data.</param>
        /// <param name="context">The destination <see cref="StreamingContext"/> for this serialization.</param>
        public override void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            base.GetObjectData(info, context);

            // Serialize command frame
            info.AddValue("command", m_command, typeof(DeviceCommand));
        }

        #endregion
    }
}